homeassistant/components/mqtt/device_trigger.py

```diff
@@ -0,0 +1,277 @@
+"""Provides device automations for MQTT."""
+import logging
+from typing import List
+
+import attr
+import voluptuous as vol
+
+from homeassistant.components import mqtt
+from homeassistant.components.automation import AutomationActionType
+import homeassistant.components.automation.mqtt as automation_mqtt
+from homeassistant.components.device_automation import TRIGGER_BASE_SCHEMA
+from homeassistant.const import CONF_DEVICE_ID, CONF_DOMAIN, CONF_PLATFORM, CONF_TYPE
+from homeassistant.core import CALLBACK_TYPE, HomeAssistant, callback
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers import config_validation as cv
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+from homeassistant.helpers.typing import ConfigType, HomeAssistantType
+
+from . import (
+    ATTR_DISCOVERY_HASH,
+    CONF_CONNECTIONS,
+    CONF_DEVICE,
+    CONF_IDENTIFIERS,
+    CONF_PAYLOAD,
+    CONF_QOS,
+    DOMAIN,
+)
+from .discovery import MQTT_DISCOVERY_UPDATED, clear_discovery_hash
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_AUTOMATION_TYPE = "automation_type"
+CONF_DISCOVERY_ID = "discovery_id"
+CONF_SUBTYPE = "subtype"
+CONF_TOPIC = "topic"
+DEFAULT_ENCODING = "utf-8"
+DEVICE = "device"
+
+MQTT_TRIGGER_BASE = {
+    # Trigger when MQTT message is received
+    CONF_PLATFORM: DEVICE,
+    CONF_DOMAIN: DOMAIN,
+}
+
+TRIGGER_SCHEMA = TRIGGER_BASE_SCHEMA.extend(
+    {
+        vol.Required(CONF_PLATFORM): DEVICE,
+        vol.Required(CONF_DOMAIN): DOMAIN,
+        vol.Required(CONF_DEVICE_ID): str,
+        vol.Required(CONF_DISCOVERY_ID): str,
+        vol.Required(CONF_TYPE): cv.string,
+        vol.Required(CONF_SUBTYPE): cv.string,
+    }
+)
+
+TRIGGER_DISCOVERY_SCHEMA = mqtt.MQTT_BASE_PLATFORM_SCHEMA.extend(
+    {
+        vol.Required(CONF_AUTOMATION_TYPE): str,
```

Comment:
this is trigger, condition or action, no ?

Reply:
That's right, and it's validated here: https://github.com/home-assistant/home-assistant/blob/95d6407c94ecbc887e53bac1953bf243b8335ef3/homeassistant/components/mqtt/device_automation.py#L20

Reply:
Then why not here ?